### DISCUSSION TITLE
Bookstack to Outline migration

### DISCUSSION THREAD

**Discussion by florian-obradovic** [outline/outline#6166] [General]
Hi Outline Community.
First of all a big thanks to the developers and the community 😘 

I just did a quick and dirty migration away from my Bookstack instance to self-hosted Outline v.0.73.1. :)
Please be nice ❤️, I'm not a coder and every script is probably ugly but did the job for me. 

I don't have the skills to write a script which respects the structure / hierarchy, page titles incl. special charcters, etc.
So my "script" just imports all pages flat into a single given collection. 

**Page Titles**
There are also "issues" with the page titles because the title is given on import by the name of the file (rename script attached).
Example: 
- Original: ZFS use disk aliases like slot1 using /etc/zfs/vdev_ id.conf (by-vdev)
- Result: zfs-use-disk-aliases-like-slot1-using-etczfsvdeidconfbyvdev 

**Images**
- they're imported ✌️ 
- full width 👎 so you have to touch them later
- image alignment is topic in Outline for itself ;)

### Requirements
- create an API token: https://outline.domain.com/settings/tokens
- Make sure Bookstack images are accessible anonymously (default) if you export via Bookstack's API
You can test by copying the image address and open in private window
https://www.bookstackapp.com/docs/admin/upload-config/#migrating-to-secure-images
STORAGE_TYPE=local

This is required for the import to work with the html files exported by the small **go** application**
https://github.com/BookStackApp/api-scripts/tree/main/go-export-page-content

The images will be imported via the source URLs on import as files, not references :)

Manually exporting the pages or whole books as single html page also works.
- [go-export-page-content](https://github.com/BookStackApp/api-scripts/tree/main/go-export-page-content)
- PowerShell (did it on macOS: brew install powershell)

### Export
- Get [go-export-page-content](https://github.com/BookStackApp/api-scripts/tree/main/go-export-page-content) compiled
- `./bookstack-export --baseurl=https://bookstack.your-domain.com --tokenid=Token123id --tokensecret=987654321secret`
- This results in a folder structure representing the Bookstack structure like:
Book1 > Page1
_______ Page2
Book2 > Page1
_______ > Chapter1 > Page1
- Do some renaming like first character uppercase, remove hyphens with space, etc.

```powershell
# Set the path to the directory
$directoryPath = "/Users/flo/Downloads/go-export-page-content/page-export/how-to-articles"

# Get all files recursively with the specified pattern
$files = Get-ChildItem -Path $directoryPath -Recurse

foreach ($file in $files) {
    # Construct the new file name with the first character in uppercase and dashes replaced with spaces
    $newBaseName = $file.BaseName -replace '-', ' '
    $newFileName = Join-Path $file.DirectoryName ($newBaseName.Substring(0, 1).ToUpper() + $newBaseName.Substring(1) + $file.Extension)

    # Display the dry run result
    Write-Output "Dry run: Rename '$($file.FullName)' to '$newFileName'"
}

# Note: The actual renaming is not performed in this script (dry run). 
# If you are satisfied with the dry run results, you can uncomment the following lines to perform the renaming:
#foreach ($file in $files) {
#    $newBaseName = $file.BaseName -replace '-', ' '
#    $newFileName = Join-Path $file.DirectoryName ($newBaseName.Substring(0, 1).ToUpper() + $newBaseName.Substring(1) + $file.Extension)
#    Rename-Item -Path $file.FullName -NewName $newFileName
#}
``` 

### Prepare Import
Define where you want to import your files to.

**List all collectionIds**
```shell
curl https://outline.domain.com/api/collections.list \
-X 'POST' \
-H 'authorization: Bearer ol_api_vVIQbOep8Jm6wr6yxdpg2rCfyG2oa81zgty86O' \
-H 'content-type: application/json' \
-H 'accept: application/json' | jq '.data[] | {name, id}'

{
  "name": "Test",
  "id": "23da4734-7c1e-4dc5-9931-cb084b704e8b"
}
{
  "name": "Knowledge Base",
  "id": "ec3eef05-8ae9-4f2c-98b8-4e8d40edb00a"
}
{
  "name": "IT Documentation",
  "id": "8dc03d67-5a9d-4a3c-8e67-50142f0a96da"
}
{
  "name": "Bookstack 2 Outline Migration",
  "id": "29423ce4-cb4b-4997-8806-4fe7af12088e"
}
``` 

**List all parentDocumentIds:**
```shell
curl https://outline.domain.com/api/documents.list \
-X 'POST' \
-H 'authorization: Bearer ol_api_vVIQbOep8Jm6wr6yxdpg2rCfyG2oa81zgty86O' \
-H 'content-type: application/json' \
-H 'accept: application/json' | jq '.data[] | {title, id}

{
  "title": "Zfs encryption bug permanent errors have been detected 0x0",
  "id": "e614b661-0b11-4a13-98c6-e12ef969ae7f"
}
{
  "title": "Winbox neighbour discovery doesnt work check firewall",
  "id": "6430be28-dea3-45a3-9796-2b78e8b1540f"
}
{
  "title": "Ubuntu 1910 2004 on hp microserver g8 stuck at purple screen no gui",
  "id": "0baf7563-7e14-40f4-8f85-3a52424fceb6"
}
```

### Final Import
```powershell
$token = "ol_api_vVIQbOep8Jm6wr6yxdpg2rCfyG2oa81zgty86O"
$collectionId = "23da4734-7c1e-4dc5-9931-cb084b704e8b"
$parentDocumentId = "879d9cb3-1c03-4d4a-af64-89a973a8d3cc"
$apiUrl = "https://outline.domain.com/api/documents.import"
$folder = "/Users/flo/Downloads/bookstack-mig/test"
# Get all files in the specified folder and its subfolders
$files = Get-ChildItem -Path $folder -File # -recurse

# Iterate through each file and run the curl command
foreach ($file in $files) {
    $filePath = $file.FullName
    $fileName = $file.Name

    # Import the files to Outline
    # With parent ID 
    # $curlCommand = "curl -X POST -H 'Authorization: Bearer $token' -H 'Content-Type: multipart/form-data' -F 'collectionId=$collectionId' -F 'parentDocumentId=$parentDocumentId' -F 'file=@$filePath' -F 'template=false' -F 'publish=true' $apiUrl | jq"
    
    # Directly under the specified collection
    $curlCommand = "curl -X POST -H 'Authorization: Bearer $token' -H 'Content-Type: multipart/form-data' -F 'collectionId=$collectionId' -F 'file=@$filePath' -F 'template=false' -F 'publish=true' $apiUrl | jq"

    # Execute the curl command
    Invoke-Expression $curlCommand
}
```

**Comment by somoit-net** [comment]
Hello!
I have been migrating from Bookstack to Outline so i created these python scripts for exporting and importing.
Happy to share them in case it helps:
https://somoit.net/sysadmin/bookstack-to-outline-migration/
https://github.com/somoit-net/bookstack_outline_migrator
